src/udp/connection_cookie.rs

```
@@ -0,0 +1,186 @@
+
+    pub(super) fn get_last_time_extent() -> SinceUnixEpochTimeExtent {
+        DefaultTimeExtentMaker::now(&COOKIE_LIFETIME.increment)
+            .unwrap()
+            .unwrap()
+            .increase(COOKIE_LIFETIME.amount)
+            .unwrap()
+    }
+
+    pub(super) fn build(remote_address: &SocketAddr, time_extent: &TimeExtent) -> Cookie {
+        let seed = DefaultSeed::get_seed();
+
+        let mut hasher = DefaultHasher::new();
+
+        remote_address.hash(&mut hasher);
+        time_extent.hash(&mut hasher);
+        seed.hash(&mut hasher);
+
+        //println!("remote_address: {remote_address:?}, time_extent: {time_extent:?}, seed: {seed:?}");
+
+        hasher.finish().to_le_bytes()
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use std::net::{IpAddr, Ipv4Addr, Ipv6Addr, SocketAddr};
+
+    use super::cookie_builder::get_last_time_extent;
+    use crate::protocol::clock::time_extent::Extent;
+    use crate::protocol::clock::{StoppedClock, StoppedTime};
+    use crate::udp::connection_cookie::{check_connection_cookie, make_connection_cookie, Cookie, COOKIE_LIFETIME};
+
+    fn make_test_socket_addr() -> SocketAddr {
+        SocketAddr::new(IpAddr::V4(Ipv4Addr::new(127, 0, 0, 1)), 8080)
+    }
+
+    fn make_test_cookie(remote_address: Option<&SocketAddr>) -> Cookie {
+        make_connection_cookie(remote_address.unwrap_or(&make_test_socket_addr()))
+    }
+
+    /// this test is strange, the default hasher seems to make different values depending if we are running stable ot nightly
+    #[test]
+    fn it_should_make_a_connection_cookie() {
+        // remote_address: 127.0.0.1:8080, time_extent: 60,
+        // seed: [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]
+
+        const ID_STABLE: Cookie = [223, 239, 248, 49, 67, 103, 97, 83];
+        const ID_NIGHTLY: Cookie = [45, 59, 50, 101, 97, 203, 48, 19];
+
+        let test_cookie = make_test_cookie(None);
```

@da2ce7 I do not know if we should allow this. I think we should throw an Error or a Cookie representing an invalid Cookie.

If `None` means we are using a loopback IP and that's OK them I would change the test name to:

`it_should_use_loopback_ip_as_default_remote_address_given_it_is_not_provided`

---

Reworked.